pkg/kubernetes/namespace.go

```diff
@@ -53,6 +55,225 @@ func CheckNamespaceExists(clientSet Interface, name string) (bool, error) {
 	return false, nil
 }
 
+// DeleteNamespace deletes the provided namespace and all resources namespaced
+// under it. It waits until all "common" (pods, deployments, services, etc.)
+// resources are deleted before returning.
+// N.B.: this method does NOT delete non-namespaced resources, e.g.:
+//       ClusterRoles or ClusterRoleBindings.
+func DeleteNamespace(clientSet Interface, namespace string) error {
+	if err := clientSet.CoreV1().Namespaces().Delete(namespace, &metav1.DeleteOptions{}); err != nil {
+		return errors.Wrapf(err, "failed to delete namespace %s", namespace)
+	}
+	logger.Info("deleting namespace %s", namespace)
+	// There may be race conditions if we return right after asynchronously
+	// deleting the namespace, hence we wait for commonly used resources.
+	// This isn't perfect in any way, but given there is no way to directly
+	// list all namespaced resources without using kubectl [1] this will
+	// hopefully do in most cases.
+	// [1]: kubectl api-resources --verbs=list --namespaced -o name \
+	//      | xargs -n 1 kubectl get --show-kind --ignore-not-found -n <namespace>
+	return waitForCommonNamespacedResourcesDeletion(clientSet, namespace)
+}
+
+func waitForCommonNamespacedResourcesDeletion(clientSet Interface, namespace string) error {
```

Comment:
FYI, Can't you get this exact effect using the `"Foreground"` DeletionPropagation policy when invoking delete on the namespace? https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#DeletionPropagation

Reply:
Maybe, will look into it. Thanks for the pointer 👍

Reply:
It may work when doing so on the namespace (still haven't tried), but for the record, it doesn't help now that we delete individual resources. 😞